README.md
# garnett

![CircleCI](https://img.shields.io/circleci/project/github/glotzerlab/garnett/master.svg)
[![RTD](https://img.shields.io/readthedocs/garnett.svg?style=flat)](https://garnett.readthedocs.io/)
[![License](https://img.shields.io/github/license/glotzerlab/garnett.svg)](https://github.com/glotzerlab/garnett/blob/master/LICENSE.txt)

## About

This is a collection of samples, parsers and writers for formats used in the Glotzer Group at the University of Michigan, Ann Arbor.

## Maintainers

* Luis Y. Rivera-Rivera, lyrivera@umich.edu
* Kelly Wang, kelwang@umich.edu
* Carl Simon Adorf, csadorf@umich.edu
* Bradley Dice, bdice@umich.edu

## Setup

To install this package with pip, execute:

    pip install git+https://github.com/glotzerlab/garnett.git#egg=garnett --user

## Documentation

**The package's documentation is available at:** [http://glotzerlab.engin.umich.edu/garnett/](http://glotzerlab.engin.umich.edu/garnett/)

To build the documentation yourself using sphinx, execute within the repository:

    cd doc
    make html
    open _build/html/index.html

## Quickstart

### Reading and writing

``` python
import garnett

# Autodetects file format for a uniform trajectory API
with garnett.read('gsdfile.gsd') as traj:
    for frame in traj:
        pos = frame.positions

# Simple conversion of trajectory formats
with garnett.read('posfile.pos') as traj:
    garnett.write(traj, 'gsdfile.gsd')
```

### Data access

Access individual frames by indexing or create sub trajectories with slicing:
```python
first_frame = traj[0]
last_frame = traj[-1]
n_th_frame = traj[n]
# and so on

sub_trajectory = traj[i:j]
```

Access properties of trajectories:
```
traj.load_arrays()
traj.positions       # MxNx3
traj.orientations    # MxNx4
traj.velocities      # MxNx3
traj.mass            # MxN
traj.charge          # MxN
traj.diameter        # MxN
traj.moment_inertia  # MxNx3
traj.angmom          # MxNx4
traj.types           # MxN

# where M=len(traj) and N=max((len(f) for f in traj))
```

Access properties of individual frames:
```
frame = traj[i]
frame.box              # 3x3 matrix (not required to be upper-triangular)
frame.types            # N
frame.positions        # Nx3
frame.orientations     # Nx4
frame.velocities       # Nx3
frame.mass             # N
frame.charge           # N
frame.diameter         # N
frame.moment_inertia   # Nx3
frame.angmom           # Nx4
frame.data             # A dictionary of lists for each attribute
frame.data_key         # A list of strings
frame.shapedef         # A ordered dictionary of instances of ShapeDefinition
```
All matrices are `numpy` arrays.

## Example use with HOOMD-blue

See the [examples directory](https://github.com/glotzerlab/garnett/tree/master/examples) for additional examples.

```
#!python
pos_reader = PosFileReader()
with open('cube.pos') as posfile:
    traj = pos_reader.read(posfile)

# Initialize from last frame
snapshot = traj[-1].make_snapshot()
system = init.read_snapshot(snapshot)

# Restore last frame
snapshot = system.take_snapshot()
traj[-1].copyto_snapshot(snapshot)

```

## Testing

To run all garnett tests, `ddt`, HOOMD-blue (`hoomd`), and `pycifrw` must be installed in the testing environments.

Execute the tests with

    python -m unittest discover tests


changelog.md
=========
Changelog
=========

The **garnett** package follows `semantic versioning <https://semver.org/>`_.

Version 0.5
===========	

[0.5.0] -- 2019-0x-xx
---------------------
Added
+++++
  - Rowan is now a dependency.
  - Add GETAR file reader/writer.
  - Add ``shape_dict`` representation to ``Shape`` classes.
  - Add support for particle properties: 
    - mass 
    - charge 
    - diameter
    - image
    - moment of inertia 
    - angular momentum
  - Add support for reading/writing shapes in GSD via HOOMD-HPMC state.
  - Add universal reader/writer with format detection.
  - Add orientable attribute to spheres.
  - Extend list of supported shape classes: 
    - ellipsoid 
    - polygon
    - spheropolygon
    - convex polyhedron
    - convex spheropolyhedron

Changed
+++++++
  - Raise ``AttributeError`` if accessing an invalid frame or trajectory property.
  - Rename several existing shape classes.
  - Improve unit test coverage.
  - Revise documentation.
  - Move shape definitions to separate module.

Deprecated
++++++++++
  - Tests for Python 2 are no longer updated (Python 2 support will be entirely dropped in a future release).

Removed
+++++++
  - Remove acceleration as supported property.
  - Remove the ``read_gsd_shape_data`` flag from GSD reader.

Version 0.4
===========

[0.4.1] -- 2017-08-23
---------------------

Fixed
+++++
  - Fix minor bug related to QR check for 2d boxes.

[0.4.0] -- 2017-06-26
---------------------

Added
+++++
  - Add readers/writers:
    - CIF reader
    - GSD writer
  - Support shape definitions:
    - spheropolyhedron
    - polyunion
    - convex polyhedron union
  - Add ``gf2pos`` script - convert to pos-file from any supported format.
  - Add shape definitions to ``GetarFileReader``.
  - Interpret the pos-file rotation key word.

Changed
+++++++
  - ``GetarFileReader`` skips records that have a non-empty group field.
  - Improve algorithm for the normalization of frames with non-standard box.
  - Various documentation updates.

Version 0.3
===========

[0.3.9] -- 2017-01-30
---------------------

Added
+++++
  - The ``GSDReader`` now reads velocities.
  - Support ``PolyV`` shape definitions.

Changed
+++++++
  - Update documentation concerning the conversion of rotations from quaternions to euler angles.
Fix
+++
  - Fix bug related to trajectory arrays when slicing the array.

[0.3.8] -- 2016-12-21
---------------------

Fixed
+++++
  - Hot fix: Negative euler angles were not read correctly in skewed boxes using the ``DCDFileReader``.

[0.3.7] -- 2016-11-07
---------------------

Added
+++++
  - Add the ``whence`` argument to the file format's seek method.

Fixed
+++++
  - Fix bug in ``DCDfilereader`` leading to incorrect box dimensions to be read for skewed boxes. Cubic or squared boxes are not affected.

[0.3.6] -- 2016-10-20
---------------------

Fixed
+++++
  - Fix quaternion to euler angle conversion example in the DCD file reader documentation.

[0.3.5] -- 2016-09-20
---------------------

Changed
+++++++
  - ``GSDHOOMDFileReader`` uses the native GSD library if installed.
  - Reduced warning verbosity.

Fixed
+++++
  - Fix bug that caused the ``GSDHOOMDFileReader`` to ignore dimensions specified in the GSD file.

[0.3.4] -- 2016-09-08
---------------------

Added
+++++
  - Support velocities in HOOMD-blue XML files.
  - Support ``SphereUnionShape`` in ``PosFileReader``.

Changed
+++++++
  - Support Pos-Files using the keyword 'box' instead of 'boxMatrix'

Fixed
+++++
  - Fix bug in ``PosFileReader`` which occured with non-standard pos-file in python 3.5
  - Fix bug, which occured when constructing frames from raw frames using box instances instead of a box matrix.

[0.3.3] -- 2016-07-19
---------------------

Fixed
+++++
  - Fix bug related to 2-dimensional systems and a box z-dimensions not equal to 1.

[0.3.2] -- 2016-07-15
---------------------

Added
+++++
  - Add ``trajectory.N``, ``trajectory.type`` and ``trajectory.type_ids`` as an alternative mode to access frame length and type information.

Fixed
+++++
  - Fix bug in ``GSDHOOMDFileReader`` when not providing template frame.

[0.3.1] -- 2016-07-08
---------------------

Changed
+++++++
  - Update the GSD hoomd module.

[0.3.0] -- 2016-07-06
---------------------

Added
+++++
  - Provide a highly optimized cythonized ``DCDFileReader``.
  - Allow trajectory data acess via coherent numpy arrays.
  - Make snapshot creation and copying HOOMD-blue 2.0 compatible.

Changed
+++++++

  - Update the GSD module.
  - Improve the ``Box`` class documentation.
  - Overall improvement of the documentation.

Fixed
+++++
  - Fix and optimize the pure-python ``DCDFileReader``.

Version 0.2
===========

[0.2.1] -- 2016-07-10
---------------------

Fixed
+++++
  - Fix an issue with injavis pos-files causing parser errors.

[0.2.0] -- 2016-04-28
---------------------

Fixed
+++++
  - Fix HOOMD-blue snapshot type issue.

Version 0.1
===========

[0.1.9] -- 2016-04-09
---------------------

Added
+++++
  - Add ``GSDHoomdFileReader``.

Fixed
+++++
  - Fix type issue in ``HoomdBlueXMLFileReader``.

[0.1.8] -- 2016-04-04
---------------------

Added
+++++
  - Add ``HoomdBlueXMLFileReader``.
  - Add ``DCDFileReader``.
  - Add ``CifFileWriter``.
  - Add ``GetarFileReader``.

Fixed
+++++
  - Fix type issue in DCD.


[0.1.6] -- 2016-01-28
---------------------

Changed
+++++++
  - Extend FileFormat API to increase file-like compatibility.

Fixed
+++++
  - Fixed ``box_matrix`` calculation.

[0.1.5] -- 2016-01-11
---------------------

Changed
+++++++
  - Frames only loaded into memory on demand.
  - Improved trajectory iteration logic.


garnett/shapes.py
# Copyright (c) 2019 The Regents of the University of Michigan
# All rights reserved.
# This software is licensed under the BSD 3-Clause License.

"""Abstract shape definitions used to read/write particle shapes."""


__all__ = [
    'FallbackShape',
    'Shape',
    'SphereShape',
    'ArrowShape',
    'SphereUnionShape',
    'PolygonShape',
    'SpheropolygonShape',
    'ConvexPolyhedronShape',
    'ConvexPolyhedronUnionShape',
    'ConvexSpheropolyhedronShape',
    'GeneralPolyhedronShape',
]


SHAPE_DEFAULT_COLOR = '005984FF'


class FallbackShape(str):
    """This shape definition class is used when no specialized Shape class can be applied.

    The fallback shape definition is a string containing the definition."""
    pass


class Shape(object):
    """Parent class of all shape objects.

    :param shape_class:
        Shape class directive, used for POS format (default: :code:`None`).
    :type shape_class:
        str
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, shape_class=None, color=None):
        self.shape_class = shape_class
        self.color = color if color else SHAPE_DEFAULT_COLOR

    def __str__(self):
        return "{} {}".format(self.shape_class, self.color)

    def __repr__(self):
        return str(self)

    def __eq__(self, other):
        return str(self) == str(other)


class SphereShape(Shape):
    """Shape class for spheres of a specified diameter.

    :param diameter:
        Diameter of the sphere.
    :type diameter:
        float
    :param orientable:
         Set to True for spheres with orientation (default: :code:`False`).
    :type orientable:
        bool
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, diameter, orientable=False, color=None):
        super(SphereShape, self).__init__(
            shape_class='sphere', color=color)
        self.diameter = diameter
        self.orientable = orientable

    def __str__(self):
        return "{} {} {}".format(self.shape_class, self.diameter, self.color)

    @property
    def shape_dict(self):
        """Shape as dictionary. Example:

            >>> SphereShape(2.0).shape_dict
            {'type': 'Sphere', 'diameter': 2.0, 'orientable': False}
        """
        return {'type': 'Sphere',
                'diameter': self.diameter,
                'orientable': self.orientable}


class ArrowShape(Shape):
    """Shape class for arrows of a specified thickness.

    :param thickness:
        Thickness of the arrow.
    :type thickness:
        float
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, thickness=0.1, color=None):
        super(ArrowShape, self).__init__(
            shape_class='arrow', color=color)
        self.thickness = thickness

    def __str__(self):
        return "{} {} {}".format(self.shape_class, self.thickness, self.color)


class SphereUnionShape(Shape):
    """Shape class for sphere unions, such as rigid bodies of many spheres.

    :param diameters:
        List of sphere diameters.
    :type diameters:
        list
    :param centers:
        List of 3D center vectors.
    :type centers:
        list
    :param colors:
        List of hexadecimal color strings in format :code:`RRGGBBAA` (default: :code:`None`).
    :type colors:
        list
    """

    def __init__(self, diameters, centers, colors=None):
        super(SphereUnionShape, self).__init__(
            shape_class='sphere_union', color='')
        self.diameters = diameters
        self.centers = centers
        self.colors = colors

    def __str__(self):
        shape_def = '{} {} '.format(self.shape_class, len(self.centers))
        for d, p, c in zip(self.diameters, self.centers, self.colors):
            shape_def += '{0} '.format(d)
            shape_def += '{0} {1} {2} '.format(*p)
            shape_def += '{0} '.format(c)

        return shape_def


class PolygonShape(Shape):
    """Shape class for polygons in a 2D plane.

    :param vertices:
        List of 2D vertex vectors.
    :type vertices:
        list
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, vertices, color=None):
        super(PolygonShape, self).__init__(
            shape_class='poly3d', color=color)
        self.vertices = vertices

    def __str__(self):
        return "{} {} {} {}".format(
            self.shape_class,
            len(self.vertices),
            ' '.join('{} {} 0'.format(v[0], v[1]) for v in self.vertices),
            self.color)

    @property
    def shape_dict(self):
        """Shape as dictionary. Example:

            >>> PolygonShape([[-0.5, -0.5], [0.5, -0.5], [0.5, 0.5]]).shape_dict
            {'type': 'Polygon', 'rounding_radius': 0,
             'vertices': [[-0.5, -0.5], [0.5, -0.5], [0.5, 0.5]]}
        """
        return {'type': 'Polygon',
                'rounding_radius': 0,
                'vertices': self.vertices}


class SpheropolygonShape(Shape):
    """Shape class for rounded polygons in a 2D plane.

    :param vertices:
        List of 2D vertex vectors.
    :type vertices:
        list
    :param rounding_radius:
        Rounding radius applied to the spheropolygon (default: 0).
    :type rounding_radius:
        float
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, vertices, rounding_radius=0, color=None):
        super(SpheropolygonShape, self).__init__(
            shape_class='spoly3d', color=color)
        self.vertices = vertices
        self.rounding_radius = rounding_radius

    def __str__(self):
        return "{} {} {} {} {}".format(
            self.shape_class,
            self.rounding_radius,
            len(self.vertices),
            ' '.join('{} {} 0'.format(v[0], v[1]) for v in self.vertices),
            self.color)

    @property
    def shape_dict(self):
        """Shape as dictionary. Example:

            >>> SpheropolygonShape([[-0.5, -0.5], [0.5, -0.5], [0.5, 0.5]], 0.1).shape_dict
            {'type': 'Polygon', 'rounding_radius': 0.1,
             'vertices': [[-0.5, -0.5], [0.5, -0.5], [0.5, 0.5]]}
        """
        return {'type': 'Polygon',
                'rounding_radius': self.rounding_radius,
                'vertices': self.vertices}


class ConvexPolyhedronShape(Shape):
    """Shape class for convex polyhedra.

    :param vertices:
        List of 3D vertex vectors.
    :type vertices:
        list
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, vertices, color=None):
        super(ConvexPolyhedronShape, self).__init__(
            shape_class='poly3d', color=color)
        self.vertices = vertices

    def __str__(self):
        return "{} {} {} {}".format(
            self.shape_class,
            len(self.vertices),
            ' '.join((str(v) for xyz in self.vertices for v in xyz)),
            self.color)

    @property
    def shape_dict(self):
        """Shape as dictionary. Example:

            >>> ConvexPolyhedronShape([[0.5, 0.5, 0.5], [0.5, -0.5, -0.5],
                                       [-0.5, 0.5, -0.5], [-0.5, -0.5, 0.5]]).shape_dict
            {'type': 'ConvexPolyhedron', 'rounding_radius': 0,
             'vertices': [[0.5, 0.5, 0.5], [0.5, -0.5, -0.5],
                          [-0.5, 0.5, -0.5], [-0.5, -0.5, 0.5]]}
        """
        return {'type': 'ConvexPolyhedron',
                'rounding_radius': 0,
                'vertices': self.vertices}


class ConvexPolyhedronUnionShape(Shape):
    """Shape class for unions of convex polyhedra.

    :param vertices:
        List of lists of 3D vertex vectors in particle coordinates (each
        polyhedron, each vertex).
    :type vertices:
        list
    :param centers:
        List of 3D polyhedra center vectors.
    :type centers:
        list
    :param orientations:
        Orientations of the polyhedra, as a list of quaternions.
    :type orientations:
        list
    :param colors:
        List of hexadecimal color strings in format :code:`RRGGBBAA` (default: :code:`None`).
    :type colors:
        list
    """

    def __init__(self, vertices, centers, orientations, colors=None):
        super(ConvexPolyhedronUnionShape, self).__init__(
            shape_class='poly3d_union', color='')
        self.vertices = vertices
        self.centers = centers
        self.orientations = orientations
        self.colors = colors

    def __str__(self):
        shape_def = '{} {} '.format(self.shape_class, len(self.centers))
        for verts, p, q, c in zip(self.vertices, self.centers, self.orientations, self.colors):
            shape_def += '{0} '.format(len(verts))
            for v in verts:
                shape_def += '{0} {1} {2} '.format(*v)
            shape_def += '{0} {1} {2} '.format(*p)
            shape_def += '{0} {1} {2} {3} '.format(*q)
            shape_def += '{0} '.format(c)

        return shape_def


class ConvexSpheropolyhedronShape(Shape):
    """Shape class for a convex polyhedron extended by a rounding radius.

    :param vertices:
        List of 3D vertex vectors.
    :type vertices:
        list
    :param rounding_radius:
        Rounding radius applied to the spheropolyhedron (default: 0).
    :type rounding_radius:
        float
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, vertices, rounding_radius=0, color=None):
        super(ConvexSpheropolyhedronShape, self).__init__(
            shape_class='spoly3d', color=color)
        self.vertices = vertices
        self.rounding_radius = rounding_radius

    def __str__(self):
        return "{} {} {} {} {}".format(
            self.shape_class,
            self.rounding_radius,
            len(self.vertices),
            ' '.join((str(v) for xyz in self.vertices for v in xyz)),
            self.color)

    @property
    def shape_dict(self):
        """Shape as dictionary. Example:

            >>> ConvexSpheropolyhedronShape([[0.5, 0.5, 0.5], [0.5, -0.5, -0.5],
                                             [-0.5, 0.5, -0.5], [-0.5, -0.5, 0.5]], 0.1).shape_dict
            {'type': 'ConvexPolyhedron', 'rounding_radius': 0.1,
             'vertices': [[0.5, 0.5, 0.5], [0.5, -0.5, -0.5],
                          [-0.5, 0.5, -0.5], [-0.5, -0.5, 0.5]]}
        """
        return {'type': 'ConvexPolyhedron',
                'rounding_radius': self.rounding_radius,
                'vertices': self.vertices}


class GeneralPolyhedronShape(Shape):
    """Shape class for general polyhedra, such as arbitrary meshes.

    :param vertices:
        List of 3D vertex vectors.
    :type vertices:
        list
    :param faces:
        List of lists of integers representing vertex indices for each face.
    :type faces:
        list
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    :param facet_colors:
        List of hexadecimal color strings in format :code:`RRGGBBAA` for each facet (default: :code:`None`).
    :type facet_colors:
        list
    """

    def __init__(self, vertices, faces, color=None, facet_colors=None):
        super(GeneralPolyhedronShape, self).__init__(
            shape_class='polyV', color=color)
        self.vertices = vertices
        self.faces = faces
        self.facet_colors = facet_colors

    def __str__(self):
        return "{} {} {} {} {} {}".format(
            self.shape_class,
            len(self.vertices),
            ' '.join((str(v) for xyz in self.vertices for v in xyz)),
            len(self.faces),
            ' '.join((str(fv) for f in self.faces for fv in [len(f)]+f)),
            self.color)

    @property
    def shape_dict(self):
        """Shape as dictionary. Example:

            >>> GeneralPolyhedronShape([[0.5, 0.5, 0.5], [0.5, -0.5, -0.5],
                                        [-0.5, 0.5, -0.5], [-0.5, -0.5, 0.5]]).shape_dict
            {'type': 'Mesh',
             'vertices': [[0.5, 0.5, 0.5], [0.5, -0.5, -0.5],
                          [-0.5, 0.5, -0.5], [-0.5, -0.5, 0.5]],
             'indices': [[1, 2, 3], [1, 2, 4], [1, 3, 4], [2, 3, 4]]}
        """
        return {'type': 'Mesh',
                'vertices': self.vertices,
                'indices': self.faces}


class EllipsoidShape(Shape):
    """Shape class for ellipsoids of with principal axes a, b, and c.

    :param a:
        Principal axis a of the ellipsoid (radius in the x direction).
    :type a:
        float
    :param b:
        Principal axis b of the ellipsoid (radius in the y direction).
    :type b:
        float
    :param c:
        Principal axis c of the ellipsoid (radius in the z direction).
    :type c:
        float
    :param color:
        Hexadecimal color string in format :code:`RRGGBBAA` (default: :code:`None`).
    :type color:
        str
    """

    def __init__(self, a, b, c, color=None):
        super(EllipsoidShape, self).__init__(
            shape_class='ellipsoid', color=color)
        self.a = a
        self.b = b
        self.c = c

    def __str__(self):
        return "{} {} {} {} {}".format(
            self.shape_class,
            self.a,
            self.b,
            self.c,
            self.color
        )

    @property
    def shape_dict(self):
        """Shape as dictionary. Example:

            >>> EllipsoidShape(7.0, 5.0, 3.0).shape_dict
            {'type': 'Ellipsoid',
            'a': 7.0,
            'b': 5.0,
            'c': 3.0}

        """
        return {'type:': 'Ellipsoid',
                'a': self.a,
                'b': self.b,
                'c': self.c}
